sycl/plugins/xrt/pi_xrt.cpp

```diff
@@ -0,0 +1,2312 @@
+}
+
+pi_result xrt_piQueueGetInfo(pi_queue command_queue, pi_queue_info param_name,
+                             size_t param_value_size, void *param_value,
+                             size_t *param_value_size_ret) {
+  sycl::detail::pi::unimplemented(__PRETTY_FUNCTION__);
+}
+
+pi_result xrt_piQueueRetain(pi_queue command_queue) {
+  incr_ref_count(command_queue);
+  return PI_SUCCESS;
+}
+
+pi_result xrt_piQueueRelease(pi_queue command_queue) {
+  decr_ref_count(command_queue);
+  return PI_SUCCESS;
+}
+
+/// All commands are executed greedily so this is a noop.
+pi_result xrt_piQueueFinish(pi_queue command_queue) {
+  assert_valid_obj(command_queue);
+  return PI_SUCCESS;
+}
+
+/// All commands are executed greedily so this is a noop.
+pi_result xrt_piQueueFlush(pi_queue command_queue) {
+  assert_valid_obj(command_queue);
+  return PI_SUCCESS;
+}
+
+/// Gets the native CUDA handle of a PI queue object
+///
+/// \param[in] queue The PI queue to get the native CUDA object of.
+/// \param[out] nativeHandle Set to the native handle of the PI queue object.
+///
+/// \return PI_SUCCESS
+pi_result xrt_piextQueueGetNativeHandle(pi_queue queue,
+                                        pi_native_handle *nativeHandle) {
+  sycl::detail::pi::unimplemented(__PRETTY_FUNCTION__);
+}
+
+/// Created a PI queue object from a CUDA queue handle.
+/// TODO: Implement this.
+/// NOTE: The created PI object takes ownership of the native handle.
+///
+/// \param[in] nativeHandle The native handle to create PI queue object from.
+/// \param[in] context is the PI context of the queue.
+/// \param[out] queue Set to the PI queue object created from native handle.
+/// \param ownNativeHandle tells if SYCL RT should assume the ownership of
+///        the native handle, if it can.
+///
+/// \return TBD
+pi_result xrt_piextQueueCreateWithNativeHandle(pi_native_handle, pi_context,
+                                               pi_queue *,
+                                               bool ownNativeHandle) {
+  sycl::detail::pi::unimplemented(__PRETTY_FUNCTION__);
+}
+
+pi_result xrt_piEnqueueMemBufferWrite(pi_queue command_queue, pi_mem buffer,
+                                      pi_bool blocking_write, size_t offset,
+                                      size_t size, const void *ptr,
+                                      pi_uint32 num_events_in_wait_list,
+                                      const pi_event *event_wait_list,
+                                      pi_event *event) {
+  assert_valid_obj(command_queue);
+  assert_valid_obj(buffer);
+  assert_valid_objs(event_wait_list, num_events_in_wait_list);
+  assert(ptr && size);
+  assert(event);
+  *event = make_ref_counted<_pi_event>().give_externally();
+
+  buffer->run_when_mapped(command_queue->device_->get(), [=] {
+    void *adjusted_ptr = ((char *)buffer->mem.mapped_ptr) + offset;
+    REPRODUCE_ADD_BUFFER(ptr, size);
+    REPRODUCE_CALL((void)std::memcpy, adjusted_ptr, ptr, size);
```

Comment:
This sounds super inefficient.

Reply:
@stsoe it would be nice to have a way to directly write some random host data to the device `bo`